techmatters/flex-plugins
# fix: Frontend spams backend with failed profiles requests [CHI-2657]

## Description
This PR fixes an existing bug, where Flex will constantly try to fetch for profiles after a failed fetch. The failure can occur for several reasons, like temporary error or lack of permissions.
To fix this issue, we consider the case where there's an `error` after loading a profile, then act accordingly:
- For 5xx errors, we add a retry mechanisms (with back-off).
- For 4xx errors, since the problem is on the client, we don't retry at all.
This behavior is abstracted in a new hook, `useLoadWithRetry`.
This PR also refactors all other profiles related "loader hooks" to use the above mentioned hook.
The idea of introducing this hook comes from an attempt to standardize how we do async operations at a state level.

This PR also fixes an existing (and mostly inoffensive) bug with the profiles list, where each time page or filters are updated, Flex performs two parallel calls to HRM, requesting for the same list of profiles. The cause of this bug is that "load profiles list hook" was being called twice in the same component tree. To fix this issue, I'm moving the components to only rely on `useProfileList` hook, and passing down an `autoload: false` property on each component that's not the actual profiles list component, in order for them to consume the state but not trigger a re-load.

A followup PR will reintroduce the changes from https://github.com/techmatters/flex-plugins/pull/2116, but reusing this same pattern. The long term ideal state is one where we defer all the "async loading" to this hook, abstracting and unifying all the retry and back-off in a single place, plus standardizing at a state (Redux) level the way we handle async-related things like `loading` and `error` states.

### Checklist
- [x] [Corresponding issue has been opened](https://tech-matters.atlassian.net/browse/CHI-2657)
- [n/a] Feature flags added
- [n/a] Strings are localized
- [n/a] Tested for chat contacts
- [n/a] Tested for call contacts

### Verification steps
- Mock `enable_client_profiles` flag and start a local Flex server. Confirm that profile load is not attempted at all on Contact Details view.
- Start an HRM local server, changing `viewProfile` permissions to `[]` (no-one).
- Start Flex server and confirm that the client stops attempting to load a profile after the first `401` response (because of lack of permissions).
- Change HRM `viewProfile` permissions to `[["everyone"]]` and restart the server, and add an intentional 5xx error in the "get profile endpoint".
- Confirm that this time, Flex will try to reload the case after each failed attempt, with an initial back-off of 1 second, doubled after each attempt, up to a maximum of 10 attempts (can reduce the max retries count to make it easier to test).
- Change HRM to use the actual logic for "get profile endpoint", in order for the whole thing to succeed next time.
- Confirm that this time, the profile is loaded and Flex works as intended.
- **PLEASE:** Regression test around all the profiles features (profiles list, profile details view, adding/updating/removing flags and sections on a profile, etc) to confirm I'm not breaking stuff.

### AFTER YOU MERGE

1. Cut a release tag using the Github workflow. Wait for it to complete and notify in the #aselo-deploys Slack channel.
2. Comment on the ticket with the release tag version AND any additional instructions required to configure an environment to test the changes.
3. Only then move the ticket into the QA column in JIRA

You are responsible for ensuring the above steps are completed. If you move a ticket into QA without advising what version to test, the QA team will assume the latest tag has the changes. If it does not, the following confusion is on you! :-P
